lte/cloud/go/services/subscriberdb_cache/worker.go

```diff
@@ -66,41 +69,120 @@ func RenewDigests(
+func renewDigestsForNetwork(
+	network string,
+	digestStore storage.DigestStore,
+	perSubDigestStore *storage.PerSubDigestStore,
+	subStore *storage.SubStore,
+) (string, []*lte_protos.SubscriberDigestWithID, error) {
+	errs := &multierror.Error{}
+	digest, prevDigest, err := updateDigest(network, digestStore)
+	if err != nil {
+		multierror.Append(errs, err)
+	}
+
+	var perSubDigests []*lte_protos.SubscriberDigestWithID
+	if errs.ErrorOrNil() == nil {
+		perSubDigests, err = updatePerSubDigests(network, perSubDigestStore)
 		if err != nil {
-			multierror.Append(errors.Wrapf(err, "set per sub digest"))
-			continue
+			multierror.Append(errs, err)
 		}
-		perSubDigestsByNetwork[network] = perSubDigests
+	}
+
+	// If all digest-related operations succeeded, and the generated digest is the same as
+	// the previous digest, then no need to update the subscriber proto cache
+	if errs.ErrorOrNil() == nil && prevDigest == digest {
+		return digest, perSubDigests, nil
+	}
 
-		digest, err := subscriberdb.GetDigest(network)
+	err = updateSubscribers(network, subStore)
+	if err != nil {
+		multierror.Append(errs, errors.Wrapf(err, "update subscriber protos for network %+v", network))
+	}
+	// TODO(wangyyt1013): add logs for updated sub protos
+	return digest, perSubDigests, errs.ErrorOrNil()
+}
+
+func updatePerSubDigests(network string, perSubDigestStore *storage.PerSubDigestStore) ([]*lte_protos.SubscriberDigestWithID, error) {
+	// The per-sub digests in store are updated en masse (collectively serialized into one blob per network);
+	// this update takes place along with every flat digest update for consistency.
+	// If an error occurs during this step, the overall last_updated_at timestamp for the network will
+	// not update, and will indicate outdated-ness instead, forcing a redo in the next loop.
+	perSubDigests, err := subscriberdb.GetPerSubscriberDigests(network)
+	if err != nil {
+		return nil, errors.Wrapf(err, "get per sub dgests to update")
+	}
+	err = perSubDigestStore.SetDigest(network, perSubDigests)
+	if err != nil {
+		return nil, errors.Wrapf(err, "set per sub digest")
+	}
+	return perSubDigests, nil
+}
+
+func updateDigest(network string, digestStore storage.DigestStore) (string, string, error) {
+	digest, err := subscriberdb.GetDigest(network)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "generate flat digest")
+	}
+	prevDigest, err := storage.GetDigest(digestStore, network)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "get previous flat digest")
+	}
+	err = digestStore.SetDigest(network, digest)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "set flat digest")
+	}
+	return digest, prevDigest, nil
+}
+
+func updateSubscribers(network string, subStore *storage.SubStore) error {
+	apnsByName, err := subscriberdb.LoadApnsByName(network)
+	if err != nil {
+		return err
+	}
+	// Ensure the temporary table is empty for use
```

Comment:
No need for this comment -- the store's inner workings are irrelevant to this fn